src/resolve/resolved-dns-transaction.c

```diff
@@ -1039,6 +1039,9 @@ void dns_transaction_process_reply(DnsTransaction *t, DnsPacket *p, bool encrypt
         if (t->state != DNS_TRANSACTION_PENDING)
                 return;
 
+        if (t->n_attempts == 1 && !IN_SET(DNS_PACKET_RCODE(p), DNS_RCODE_SUCCESS, DNS_RCODE_NXDOMAIN))
```

Comment:
I have added the condition 'attempt == 1' to avoid increasing the counter for each negative response (including retry attempts) from the upstream server. This helps provide a clear picture of the number of failed queries and the number of queries served with stale data.

Reply:
please add a comment that explains it, as it's non-obvious

Reply:
Added comment